src/scores/probability/crps_impl.py

```diff
@@ -872,3 +874,217 @@ def crps_for_ensemble(
+        weights: Weights for calculating a weighted mean of individual scores. Note that
+            these weights are different to threshold weighting which is done by decision
+            threshold.
+
+    Returns:
+        xarray object of twCRPS values.
+
+    Raises:
+        ValueError: when ``method`` is not one of "ecdf" or "fair".
+
+    Notes:
+        Chaining functions can be created to vary the weights across given dimensions
+        such as varying the weights by climatological values.
+
+    References:
+        Allen, S., Ginsbourger, D., & Ziegel, J. (2023). Evaluating forecasts for high-impact
+        events using transformed kernel scores. SIAM/ASA Journal on Uncertainty
+        Quantification, 11(3), 906-940. https://doi.org/10.1137/22M1532184
+
+    See also:ß
+        :py:func:`scores.probability.crps_for_ensemble`
+        :py:func:`scores.probability.tail_twcrps_for_ensemble`
+        :py:func:`scores.probability.crps_cdf`
+
+
+    Examples:
+        Calculate the twCRPS for an ensemble of forecasts where the chaining function is 
+        derived from a weight function that assigns a weight of 1 to thresholds above
+        0.5 and a weight of 0 to thresholds below 0.5.
+
+        >>> import numpy as np
+        >>> import xarray as xr
+        >>> from scores.probability import twcrps_for_ensemble
+        >>> fcst = xr.DataArray(np.random.rand(10, 10), dims=['time', 'ensemble'])
+        >>> obs = xr.DataArray(np.random.rand(10), dims=['time'])
+        >>> twcrps_for_ensemble(fcst, obs, 'ensemble', lambda x: np.maximum(x, 0.5))
+
+    """
+
+    fcst = chaining_func(fcst)
+    obs = chaining_func(obs)
+    result = crps_for_ensemble(
+        fcst,
+        obs,
+        ensemble_member_dim,
+        method=method,
+        reduce_dims=reduce_dims,
+        preserve_dims=preserve_dims,
+        weights=weights,
+    )
+    return result
+
+
+def tail_twcrps_for_ensemble(
+    fcst: XarrayLike,
+    obs: XarrayLike,
+    ensemble_member_dim: str,
+    threshold: Union[XarrayLike, float],
+    *,  # Force keywords arguments to be keyword-only
+    tail: Literal["upper", "lower"] = "upper",
+    method: Literal["ecdf", "fair"] = "ecdf",
+    reduce_dims: Optional[Sequence[str]] = None,
+    preserve_dims: Optional[Sequence[str]] = None,
+    weights: Optional[XarrayLike] = None,
+) -> XarrayLike:
+    """
+    Calculates the threshold weighted continuous ranked probability score (twCRPS)
+    weighted for a tail of the distribution from ensemble input.
+
+    A threshold weight of 1 is assigned for values of the tail and a threshold weight of 0 otherwise.
+    The threshold value of where the tail begins is specified by the ``threshold`` argument.
+    The ``tail`` argument specifies whether the tail is the upper or lower tail.
+    For example, if we only care about values above 40 degrees C, we can set ``threshold=40`` and ``tail="upper"``.
```

Comment:
Given this, would we want less-than, less-than-or-equal-to, greater-than and greater-than-or-equal-to?

Reply:
Even if only supporting "upper" and "lower", the docs need to be clear about whether the threshold value is inclusive or exclusive. If it's not too much work, I think it would be more useful to support the four options and name them to match, perhaps as I have included in my first comment.

Reply:
I've updated the docstring to be explicit that the tail does not include the threshold (for now). If I add both options, I'll have to come back another day to add unit tests and do the implementation

Reply:
Actually, I need to double check that's right.

Reply:
Just an update - I now have a way forward with this. I'll make updates in the coming days

Reply:
@tennlee , I'm happy with what Nick has done here regarding whether or not the threshold is included. The answer will be the same regardless.

By analogy, it is like writing a function that returns the area under the curve `y=x**2` for `0 < x < 2` (thresholds not included) and then implementing a separate option for `0 <= x <= 2` (thresholds included). The answers will be the same.

Reply:
If the result is mathematically the same regardless of the fencepost, then it doesn't really matter the same way, although I would suggest putting a note to that effect in the docstring for people to whom it isn't immediately obvious (assuming some readers may be using it without being across the semantics). There is some argument (although perhaps not a strong one), to allow both of > and >= so that if there's something like a service directive at play that users can align to it even if the result is unchanged mathematically. But, if the maths is the same either way, then it's not about the result but rather about nicenessof API design, and much less significant, so I'll leave it to Nick to make the call on this one after reflecting on all the comments.

Reply:
Thanks for the analogy Rob!

I'd prefer to discuss this in the tutorial notebook rather than the docstring if that's okay.

I think adding a `>` or` >=` arg would could confusion since it would be a dummy argument that did nothing so I would prefer to keep it as is. If it matters to people, they can just use the `tw_crps` function with their chosen `chaining_func` to match the weighting function.

Reply:
That's totally fine. Thanks for the discussion.